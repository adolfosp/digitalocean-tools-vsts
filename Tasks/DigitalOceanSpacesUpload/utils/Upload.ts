import AWS from 'aws-sdk'
import * as fs from 'fs'
import { isEmpty } from 'lodash'
import * as path from 'path'
import tl from './tl'
import { Spaces } from '../common/Spaces'
import { Parameters } from './Parameters'
import { findFiles, getMimeTypes } from './utils'
const { default: PQueue } = require('p-queue')
import prettyBytes = require('pretty-bytes')

interface NormalizePathParameters {
  filePath: string
  digitalSourceFolder?: string
  digitalFlattenFolders: boolean
  digitalTargetFolder?: string
}

interface UploadFileParameters {
  filePath: string
  targetPath: string
  digitalAcl: string
  digitalBucket: string
  contentType: string
}

export class Upload extends Spaces<Parameters> {
  constructor(params: Parameters) {
    super(params)
  }

  public async init(): Promise<void> {
    console.log(
      tl.loc(
        'UploadingFiles',
        this.params.digitalSourceFolder,
        this.params.digitalTargetFolder
          ? this.params.digitalTargetFolder
          : 'root',
        this.params.digitalBucket
      )
    )

    const files: string[] = findFiles(this.params)

    if (isEmpty(files)) {
      console.log(tl.loc('FileNotFound', this.params.digitalSourceFolder))
      return
    }

    const uploadQueue = new PQueue({ concurrency: 4 })

    for (const filePath of files) {
      const targetPath = this.normalizeKeyPath({ ...this.params, filePath })

      const contentType = getMimeTypes({
        filePath,
        digitalContentType: this.params.digitalContentType,
      })

      console.log(tl.loc('UploadingFile', filePath, targetPath, contentType))

      uploadQueue.add(async () => {
        try {
          await this.uploadFile({
            filePath,
            targetPath,
            digitalAcl: this.params.digitalAcl,
            digitalBucket: this.params.digitalBucket,
            contentType,
          })
          console.log(tl.loc('FileUploadCompleted', filePath, targetPath))
        } catch (error) {
          console.error(tl.loc('FileUploadFailed'), error)
          throw error
        }
      })
    }

    await uploadQueue.onIdle()

    console.log(tl.loc('TaskCompleted'))
  }

  async uploadFile(
    params: UploadFileParameters
  ): Promise<AWS.S3.ManagedUpload.SendData> {
    const sendParams: AWS.S3.PutObjectRequest = {
      Bucket: params.digitalBucket,
      ACL: params.digitalAcl,
      Key: params.targetPath,
      Body: fs.createReadStream(params.filePath),
      ContentType: params.contentType,
    }

    const request: AWS.S3.ManagedUpload = this.s3Connection.upload(sendParams)

    request.on('httpUploadProgress', (progress) => {
      console.log(
        tl.loc(
          'FileUploadProgress',
          prettyBytes(progress.loaded),
          prettyBytes(progress.total),
          Math.floor((progress.loaded / progress.total) * 100).toFixed(1)
        )
      )
    })

    return request.promise()
  }

  normalizeKeyPath(parameters: NormalizePathParameters): string {
    let relativePath = parameters.filePath.substring(
      parameters.digitalSourceFolder.length
    )

    if (relativePath.startsWith(path.sep)) {
      relativePath = relativePath.substr(1)
    }

    let targetPath

    if (parameters.digitalFlattenFolders) {
      const flatFileName = path.basename(parameters.filePath)
      targetPath = parameters.digitalTargetFolder
        ? path.join(parameters.digitalTargetFolder, flatFileName)
        : flatFileName
    } else {
      targetPath = parameters.digitalTargetFolder
        ? path.join(parameters.digitalTargetFolder, relativePath)
        : relativePath
    }

    return targetPath.replace(/\\/g, '/')
  }
}
