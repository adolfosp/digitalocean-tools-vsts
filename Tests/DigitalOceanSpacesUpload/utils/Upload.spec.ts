import { Upload } from '@DOSUpload/utils/Upload.ts'
import { EventEmitter } from 'events'
// tslint:disable-next-line: no-var-requires
const AWS = require('aws-sdk')

interface MySelf extends EventEmitter {
  promise?: (data: string) => Promise<unknown>
}

const spyLog = jest.spyOn(console, 'log')

describe('DOSUpload', () => {
  const baseParameters = {
    digitalSourceFolder: './Tests/fixtures/',
    digitalFlattenFolders: false,
    digitalGlobExpressions: ['**'],
    digitalAcl: 'test',
    digitalEndpoint: {
      parameters: { username: 'test', password: 'test' },
      scheme: 'test',
    },
    digitalRegion: 'test',
    digitalBucket: 'test',
    digitalCredentials: 'test',
  }

  afterEach(() => {
    spyLog.mockClear()
    AWS.clearAllMocks()
  })

  test('should upload file successfully', async () => {
    const uploadFiles: jest.Mock = AWS.spyOn('S3', 'upload').mockImplementation(
      (params: any) => {
        const self: MySelf = new EventEmitter()
        self.promise = (data: string) =>
          new Promise((resolve) => setTimeout(() => resolve(data), 0))
        setTimeout(() => {
          self.emit('httpUploadProgress', { loaded: 1337, total: 2337 })
          self.emit('httpUploadProgress', { loaded: 2337, total: 2337 })
        }, 0)
        return self
      }
    )

    const upload = new Upload(baseParameters)

    const normalizePaths = jest.spyOn(upload, 'normalizeKeyPath')

    await upload.init()

    expect(uploadFiles.mock.calls[0][0].ACL).toEqual('test')
    expect(uploadFiles.mock.calls[0][0].Bucket).toEqual('test')
    expect(uploadFiles.mock.calls[0][0].ContentType).toEqual('text/plain')
    expect(uploadFiles.mock.calls[0][0].Key).toEqual('file-v1.0.1.txt')

    expect(uploadFiles.mock.calls[1][0].ACL).toEqual('test')
    expect(uploadFiles.mock.calls[1][0].Bucket).toEqual('test')
    expect(uploadFiles.mock.calls[1][0].ContentType).toEqual('text/plain')
    expect(uploadFiles.mock.calls[1][0].Key).toEqual('file1-v1.2.1.txt')

    expect(uploadFiles.mock.calls[2][0].ACL).toEqual('test')
    expect(uploadFiles.mock.calls[2][0].Bucket).toEqual('test')
    expect(uploadFiles.mock.calls[2][0].ContentType).toEqual(
      'application/octet-stream'
    )
    expect(uploadFiles.mock.calls[2][0].Key).toEqual('file2-v1.3.1.json')

    expect(normalizePaths).toHaveBeenCalledTimes(3)

    const baseNormalizePathsReturn = {
      digitalAcl: 'test',
      digitalBucket: 'test',
      digitalCredentials: 'test',
      digitalEndpoint: {
        parameters: { password: 'test', username: 'test' },
        scheme: 'test',
      },
      digitalFlattenFolders: false,
      digitalGlobExpressions: ['**'],
      digitalRegion: 'test',
      digitalSourceFolder: 'Tests/fixtures/',
    }

    expect(normalizePaths).toHaveBeenNthCalledWith(1, {
      ...baseNormalizePathsReturn,
      filePath: 'Tests/fixtures/file-v1.0.1.txt',
    })
    expect(normalizePaths).toHaveBeenNthCalledWith(2, {
      ...baseNormalizePathsReturn,
      filePath: 'Tests/fixtures/file1-v1.2.1.txt',
    })
    expect(normalizePaths).toHaveBeenNthCalledWith(3, {
      ...baseNormalizePathsReturn,
      filePath: 'Tests/fixtures/file2-v1.3.1.json',
    })
    expect(normalizePaths.mock.results).toEqual([
      { type: 'return', value: 'file-v1.0.1.txt' },
      { type: 'return', value: 'file1-v1.2.1.txt' },
      { type: 'return', value: 'file2-v1.3.1.json' },
    ])

    expect(spyLog.mock.calls[3][0]).toEqual(
      'Upload progress is 1.34 kB of 2.34 kB - 57%'
    )
    expect(spyLog).toHaveBeenCalledTimes(15)
  })

  test('should return "file not found" log and stop operation without error', async () => {
    const upload = new Upload({
      ...baseParameters,
      digitalSourceFolder: './Tests/fixtures/nowhere',
      digitalTargetFolder: '/pathDOS',
    })
    await upload.init()

    expect(spyLog.mock.calls[2][0]).toEqual(
      'No files found at Tests/fixtures/nowhere'
    )
  })

  describe('normalizeKeyPath', () => {
    test('should return normalized path with flatten folders', () => {
      const upload = new Upload(baseParameters)

      const normalizeKeyPathResult = upload.normalizeKeyPath({
        filePath: './Tests/fixtures/file-v1.0.1.txt',
        digitalSourceFolder: './Tests/',
        digitalFlattenFolders: true,
      })

      expect(normalizeKeyPathResult).toEqual('file-v1.0.1.txt')
    })
    test('should return normalized path with flatten folders and remove extra path.sep', () => {
      const upload = new Upload(baseParameters)

      const normalizeKeyPathResult = upload.normalizeKeyPath({
        filePath: './Tests/fixtures/file-v1.0.1.txt',
        digitalSourceFolder: './Tests',
        digitalFlattenFolders: true,
      })

      expect(normalizeKeyPathResult).toEqual('file-v1.0.1.txt')
    })
    test('should return normalized path with flatten folders and set correctly target folder', () => {
      const upload = new Upload(baseParameters)

      const normalizeKeyPathResult = upload.normalizeKeyPath({
        filePath: './Tests/fixtures/file-v1.0.1.txt',
        digitalSourceFolder: './Tests/',
        digitalFlattenFolders: true,
        digitalTargetFolder: 'pathDOS/',
      })

      expect(normalizeKeyPathResult).toEqual('pathDOS/file-v1.0.1.txt')
    })
    test('should return normalized path without flatten folders and set correctly target folder', () => {
      const upload = new Upload(baseParameters)

      const normalizeKeyPathResult = upload.normalizeKeyPath({
        filePath: './Tests/fixtures/file-v1.0.1.txt',
        digitalSourceFolder: './Tests/',
        digitalFlattenFolders: false,
        digitalTargetFolder: 'pathDOS/',
      })

      expect(normalizeKeyPathResult).toEqual('pathDOS/fixtures/file-v1.0.1.txt')
    })
  })
})
