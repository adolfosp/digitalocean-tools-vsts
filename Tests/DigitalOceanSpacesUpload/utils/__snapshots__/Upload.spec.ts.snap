// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOSUpload should upload file successfully 1`] = `
Array [
  Array [
    "All uploads to Space completed",
  ],
  Array [
    "Completed upload of Tests/fixtures/file-v1.0.1.txt to file-v1.0.1.txt",
  ],
  Array [
    "Completed upload of Tests/fixtures/file1-v1.2.1.txt to file1-v1.2.1.txt",
  ],
  Array [
    "Completed upload of Tests/fixtures/file2-v1.3.1.json to file2-v1.3.1.json",
  ],
  Array [
    "Searching ./Tests/fixtures/ for files to upload",
  ],
  Array [
    "Upload progress is 1.34 kB of 2.34 kB - 57%",
  ],
  Array [
    "Upload progress is 1.34 kB of 2.34 kB - 57%",
  ],
  Array [
    "Upload progress is 1.34 kB of 2.34 kB - 57%",
  ],
  Array [
    "Upload progress is 2.34 kB of 2.34 kB - 100%",
  ],
  Array [
    "Upload progress is 2.34 kB of 2.34 kB - 100%",
  ],
  Array [
    "Upload progress is 2.34 kB of 2.34 kB - 100%",
  ],
  Array [
    "Uploading files from ./Tests/fixtures/ to root in bucket test",
  ],
  Array [
    "Uploading matched file Tests/fixtures/file-v1.0.1.txt in file-v1.0.1.txt text/plain",
  ],
  Array [
    "Uploading matched file Tests/fixtures/file1-v1.2.1.txt in file1-v1.2.1.txt text/plain",
  ],
  Array [
    "Uploading matched file Tests/fixtures/file2-v1.3.1.json in file2-v1.3.1.json application/octet-stream",
  ],
]
`;
